tests/test_all_gather.py

```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+
+import torch
+import poptorch
+import poptorch_experimental_addons as pea
+import random
+import numpy as np
+import einops
+
+
+def set_seed(seed: int):
+    torch.manual_seed(seed)
+    random.seed(seed)
+    np.random.seed(seed)
+
+
+class _ReplicatedAllGather(torch.nn.Module):
+    def __init__(self, replication_factor: int = 1):
+        super().__init__()
+        self.replication_factor = replication_factor
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return pea.collectives.replicated_all_gather(x, self.replication_factor)
+
+
+def test_all_gather():
+    set_seed(112358)
+    x = torch.randn(32, 24)
+    options = poptorch.Options()
+    dp = 16
+    options.replicationFactor(dp)
+    model = _ReplicatedAllGather(dp)
+    model = poptorch.inferenceModel(model, options)
+    y = model(x)
```

Comment:
I wouldn't usually ask for a grad test of the core op that you didn't implement in this codebase - but what you've said previously makes me very suspicious, so maybe worthwhile?

Reply:
I can do, but I wrote in README that this PR is to support for forward only. Will address grads in a future PR

Reply:
As discussed - if it's possible to either assert inference mode or implement grads early, I'd be keen!

Reply:
It doesn't appear to be possible to grab this kind of poptorch session information from inside a python function. I'll write a grad test.